libcnb-data/src/buildpackage.rs

```diff
@@ -0,0 +1,264 @@
+///
+/// # Example
+/// ```
+/// use libcnb_data::buildpackage::Buildpackage;
+///
+/// let toml_str = r#"
+/// [buildpack]
+/// uri = "."
+///
+/// [[dependencies]]
+/// uri = "libcnb:buildpack_id"
+///
+/// [[dependencies]]
+/// uri = "../relative/path"
+///
+/// [[dependencies]]
+/// uri = "/absolute/path"
+///
+/// [[dependencies]]
+/// uri = "docker://docker.io/heroku/procfile-cnb:2.0.0"
+///
+/// [platform]
+/// os = "windows"
+/// "#;
+///
+/// let buildpackage = toml::from_str::<Buildpackage>(toml_str).unwrap();
+/// ```
+#[derive(Debug, Deserialize, Serialize, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct Buildpackage {
+    /// The buildpack to package.
+    pub buildpack: BuildpackageBuildpackReference,
+
+    /// A set of dependent buildpack locations, for packaging a meta-buildpack. Each dependent buildpack location must correspond to an order group within the meta-buildpack being packaged.
+    #[serde(default)]
+    pub dependencies: Vec<BuildpackageDependency>,
+
+    /// The expected runtime environment for the buildpackage.
+    #[serde(default)]
+    pub platform: Platform,
+}
+
+impl Default for Buildpackage {
+    fn default() -> Self {
+        Buildpackage {
+            buildpack: BuildpackageBuildpackReference::try_from(".")
+                .expect("a package.toml with buildpack.uri=\".\" should be valid"),
+            dependencies: vec![],
+            platform: Platform::default(),
+        }
+    }
+}
+
+/// The buildpack to package.
+#[derive(Debug, Deserialize, Serialize, Eq, PartialEq, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct BuildpackageBuildpackReference {
+    /// A URL or path to an archive, or a path to a directory.
+    /// If the `uri` field is a relative path it will be relative to the `package.toml` file.
+    #[serde(deserialize_with = "deserialize_uri_reference")]
+    #[serde(serialize_with = "serialize_uri_reference")]
+    pub uri: URIReference<'static>,
+}
+
+#[derive(Debug)]
+pub enum BuildpackageBuildpackError {
+    InvalidUri(String),
+}
+
+impl TryFrom<&str> for BuildpackageBuildpackReference {
+    type Error = BuildpackageBuildpackError;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        try_uri_from_str(value)
+            .map(|uri| BuildpackageBuildpackReference { uri })
+            .map_err(|_| BuildpackageBuildpackError::InvalidUri(value.to_string()))
+    }
+}
+
+/// A dependent buildpack location for packaging a meta-buildpack.
+#[derive(Debug, Deserialize, Serialize, Eq, PartialEq, Clone)]
+#[serde(deny_unknown_fields)]
+pub struct BuildpackageDependency {
+    /// A URL or path to an archive, a packaged buildpack (saved as a .cnb file), or a directory.
+    /// If the `uri` field is a relative path it will be relative to the `package.toml` file.
+    #[serde(deserialize_with = "deserialize_uri_reference")]
+    #[serde(serialize_with = "serialize_uri_reference")]
```

Comment:
@edmorley will add the comment.  the reason why we can't use the `uriparse` Serde support is because it expects a value that contains Uri-related fields (e.g.; path, host, scheme, etc.) but the value we have in the TOML file is just a plain string.